### ISSUE TITLE
training?

### ISSUE THREAD

**Issue by dflateau** [pancakewaffles/metasrgan-tutorial#1]
Is there any code or guidence for training these models? Thanks so much - great code and comments.

**Comment by pancakewaffles** [comment]
I just uploaded Tutorial_Super_Resolution_full.ipynb. It is divided into three parts, Part I: Traditional Interpolation like Bicubic Interpolation, Part II: EDSR, and Part III: Meta-SRGAN. Part II and Part III includes the training code for training their respective networks (EDSR and Meta-SRGAN).

It contains additional stuff like training EDSR and also some primer on Bicubic Interpolation but Part III: Meta-SRGAN walks you through training both the discriminator and the generator, at least the way I did it. It describes the loss functions I used, and also the training loop (e.g. grabbing data from the data loader, processing it into a batch, putting the input matrix into the Weight Prediction Network etc).

Feel free to ignore on those additional stuff and just focus on Part III: Meta-SRGAN. That is what you are looking for.

Couple of things to note:
In the full tutorial I said that we can use a v of all ones. Don't use that. I actually got worse results. Stick with Hu's implementation. Copied it here for convenience:

```
def create_v(inH, inW, r):
  scale_int = int(math.ceil(r))
  oversizedH = scale_int * inH
  oversizedW = scale_int * inW
  srH = int(inH * r)
  srW = int(inW * r)

  h = torch.arange(0,srH,1).float().div(r) - torch.arange(0,srH,1).float().div(r).floor()
  sr_to_lr_map = torch.arange(0,srH,1).float().div(r).floor().int()
  result_h = torch.ones(inH,scale_int)

  current_influence = 0
  prev_lr = 0
  for sr, lr in enumerate(sr_to_lr_map):
    if lr == prev_lr:
      result_h[lr,current_influence] = h[sr]
      prev_lr = lr
      current_influence += 1
    else:
      current_influence = 0
      result_h[lr,current_influence] = h[sr]
      current_influence += 1
      prev_lr = lr

  w = torch.arange(0,srW,1).float().div(r) - torch.arange(0,srW,1).float().div(r).floor()
  sr_to_lr_map = torch.arange(0,srW,1).float().div(r).floor().int()
  result_w = torch.ones(inW,scale_int)

  current_influence = 0
  prev_lr = 0
  for sr, lr in enumerate(sr_to_lr_map):
    if lr == prev_lr:
      result_w[lr,current_influence] = w[sr]
      prev_lr = lr
      current_influence += 1
    else:
      current_influence = 0
      result_w[lr,current_influence] = w[sr]
      current_influence += 1
      prev_lr = lr

  result_h = result_h.unsqueeze(2)
  result_h = torch.cat([result_h]*oversizedW,2).view(-1,oversizedW,1) #scale_int *inW

  result_w = result_w.unsqueeze(0)
  result_w = torch.cat([result_w]*oversizedH,0).view(-1,oversizedW,1)
  
  result = torch.cat([result_h,result_w],2).view(-1,2) # Left join

  scale_information = torch.empty(result.shape[0], 1)
  nn.init.constant_(scale_information, 1.0/r)
  result = torch.cat([result,scale_information],1)

  return result
```
  
  ^Basically in the Tutorial_Super_Resolution_full.ipynb just replace create_v() in the full tutorial with the above.

I also used 5 epochs for illustration's purposes. You can definitely train it for longer. For training details I did something similar to what I did in my paper https://arxiv.org/pdf/2004.02086.pdf, so you can use that for reference. Here is the relevant section:
![image](https://user-images.githubusercontent.com/24697599/128427056-daf11953-5374-4f90-b550-b4b3bcd75972.png)

**Comment by dflateau** [comment]
Thanks so much for your quick and thorough response! Will take a look!

**Comment by dflateau** [comment]
Got a chance to try it out - fantastic doument, BTW, very detailed and helpful. I also replaced the v vector with the code above.

When I went to train the metaSRGAN, it threw this error:

```

for epoch in range(num_epochs):
  trainSRGAN(epoch)
  
  
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-26-b9cef5f27678> in <module>
      1 for epoch in range(num_epochs):
----> 2   trainSRGAN(epoch)

<ipython-input-16-e8559c12625e> in trainSRGAN(epoch)
     52     g_loss = l1_loss + 0.001 * adversarial_loss
     53 
---> 54     g_loss.backward()
     55     optimiserG.step()
     56 

C:\Anaconda3\envs\metasrgan\lib\site-packages\torch\_tensor.py in backward(self, gradient, retain_graph, create_graph, inputs)
    253                 create_graph=create_graph,
    254                 inputs=inputs)
--> 255         torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
    256 
    257     def register_hook(self, hook):

C:\Anaconda3\envs\metasrgan\lib\site-packages\torch\autograd\__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables, inputs)
    147     Variable._execution_engine.run_backward(
    148         tensors, grad_tensors_, retain_graph, create_graph, inputs,
--> 149         allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
    150 
    151 

RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation: [torch.FloatTensor [1, 1024, 1, 1]] is at version 2; expected version 1 instead. Hint: enable anomaly detection to find the operation that failed to compute its gradient, with torch.autograd.set_detect_anomaly(True).
```

Any ideas as to what this is about? Again, thanks so much!

**Comment by pancakewaffles** [comment]
Yes! I actually had this problem. PyTorch changed version and it's a bit complainy about the order of stepping through the optimiser and calling backward. It's because I did optimiserD.step() first. Basically the change is really simple - just reorder the steps. Here is the training code I used for training Meta-SRGAN for a different project, last updated 6 months ago:

```
def trainSRGAN(epoch):
  tk1 = tqdm_notebook(enumerate(training_data_loader),total=len(training_data_loader),leave=False)
  epoch_generator_score = 0
  for i, (inputs,targets) in tk1:
    inputs, targets, train_scale = process_batch((inputs,targets))
    N,C,inH,inW = inputs.shape
    netG.update_scale(train_scale)
    inputs = inputs.to(device)
    targets = targets.to(device)
    
    v = create_v(inH,inW,train_scale).to(device)
    inputs = gfft(inputs)
    ############################
    # (1) Update D network: maximize D(x)-1-D(G(z))
    ###########################
    fake_imgs = netG(inputs,v)

    netD.zero_grad()
    real_out = netD(targets).mean()
    fake_out = netD(fake_imgs).mean()
    d_loss = 1 - real_out + fake_out
    d_loss.backward(retain_graph=True) # What's retain_graph? Allows us to keep the computational graph so that we can do backprop through it.

    ############################
    # (2) Update G network: minimize 1-D(G(z)) + Content Loss + L1 Loss
    ###########################
    netG.zero_grad()
    # Adversarial Loss
    adversarial_loss = torch.mean(1-fake_out)
    # Content Loss
    content_loss = nn.MSELoss()(netContent(fake_imgs),netContent(targets))
    # Texture Loss
    '''
    textures_fake = netTexture(fake_imgs)
    textures_target = netTexture(targets)
    texture_loss = 0
    for m1, m2 in zip(textures_fake,textures_target):
      texture_loss += nn.L1Loss()(gram_matrix(m1), gram_matrix(m2))
      '''
    # L1 Loss
    l1_loss = nn.L1Loss()(fake_imgs,targets)
    g_loss = l1_loss + 0.006 * content_loss #+ 0.001 * adversarial_loss #+ 0.001 * texture_loss
    g_loss.backward()

    optimiserD.step()
    optimiserG.step()

    fake_imgs = netG(inputs,v)
    #fake_imgs = fake_imgs.mul(pixel_range).clamp(0, 255).round().div(pixel_range)
    
    fake_out = netD(fake_imgs).mean()

    epoch_generator_score += fake_out.item()

```

The only change I think is that I moved the optimiserD.step() to the bottom alongside optimiserG.step(). 

Let me know if this works.

**Comment by dflateau** [comment]
Thaanks so much again for your quick reply! It seems like the code above calls a new routine,

`inputs = gfft(inputs)`

that isn't found in the old code. Any idea what this might be?  Thanks!

**Comment by pancakewaffles** [comment]
Yes it is related to a different project. I was training Meta-SRGAN for a different project. Just put the optimiserD.step() down next to optimiserG.step().

**Comment by dflateau** [comment]
Thanks again!  That change did it get off and training. It looks like this model is set for greyscale, so to invoke an RGB model,  I changed:

```
netG = Generator(UPSCALE_FACTOR,in_channels = 3, out_channels = 3).to(device)
netD = Discriminator(in_channels = 3).to(device)
```

but that threw

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-55-b9cef5f27678> in <module>
      1 for epoch in range(num_epochs):
----> 2   trainSRGAN(epoch)

<ipython-input-54-86b89bd8d797> in trainSRGAN(epoch)
     99         # (1) Update D network: maximize D(x)-1-D(G(z))
    100         ###########################
--> 101         fake_imgs = netG(inputs,v)
    102         '''
    103         To really train a GAN well, first we should give the generator some advantage.

C:\Anaconda3\envs\metasrgan\lib\site-packages\torch\nn\modules\module.py in _call_impl(self, *input, **kwargs)
   1049         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1050                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1051             return forward_call(*input, **kwargs)
   1052         # Do not call functions when jit is used
   1053         full_backward_hooks, non_full_backward_hooks = [], []

<ipython-input-34-45aa81811687> in forward(self, x, v)
     50       :return: Super-resolved image, of size (batch_size, out_channels, upscale_factor*input_H, upscale_factor*input_W)
     51       """
---> 52       block1 = self.block1(x)
     53       block2 = self.block2(block1)
     54       block3 = self.block3(block2)

C:\Anaconda3\envs\metasrgan\lib\site-packages\torch\nn\modules\module.py in _call_impl(self, *input, **kwargs)
   1049         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1050                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1051             return forward_call(*input, **kwargs)
   1052         # Do not call functions when jit is used
   1053         full_backward_hooks, non_full_backward_hooks = [], []

C:\Anaconda3\envs\metasrgan\lib\site-packages\torch\nn\modules\container.py in forward(self, input)
    137     def forward(self, input):
    138         for module in self:
--> 139             input = module(input)
    140         return input
    141 

C:\Anaconda3\envs\metasrgan\lib\site-packages\torch\nn\modules\module.py in _call_impl(self, *input, **kwargs)
   1049         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1050                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1051             return forward_call(*input, **kwargs)
   1052         # Do not call functions when jit is used
   1053         full_backward_hooks, non_full_backward_hooks = [], []

C:\Anaconda3\envs\metasrgan\lib\site-packages\torch\nn\modules\conv.py in forward(self, input)
    441 
    442     def forward(self, input: Tensor) -> Tensor:
--> 443         return self._conv_forward(input, self.weight, self.bias)
    444 
    445 class Conv3d(_ConvNd):

C:\Anaconda3\envs\metasrgan\lib\site-packages\torch\nn\modules\conv.py in _conv_forward(self, input, weight, bias)
    438                             _pair(0), self.dilation, self.groups)
    439         return F.conv2d(input, weight, bias, self.stride,
--> 440                         self.padding, self.dilation, self.groups)
    441 
    442     def forward(self, input: Tensor) -> Tensor:

RuntimeError: Given groups=1, weight of size [64, 3, 9, 9], expected input[4, 1, 64, 64] to have 3 channels, but got 1 channels instead
```

Does this have to do with the way v is created? Again, much thank for all your help.

**Comment by pancakewaffles** [comment]
No this has nothing to do with the way v is created. v has no information about the channels. It's your input generated by your Dataset class.

So, the error has to do with 
```
RuntimeError: Given groups=1, weight of size [64, 3, 9, 9], expected input[4, 1, 64, 64] to have 3 channels, but got 1 channels instead
```
The weights are 3 dimensional i.e. RGB (as you can see there is a number 3 there. but the input is of 1 dimensional (you can see input has a 1 there). 

So go to your Dataset class for both your Train and Test data, go to the __getitem__ function.
```
class TrainDatasetFromFolder(Dataset):
  ... blah blah ...
    
  def __getitem__(self,index):
    """
    Gets a pair of input and target (ground truth) images.
    :param index: the index of the image to obtain the pair of input and target images from. 
    The code index = index // self.num_patches provides a space-efficient way to
    obtain multiple patches per image without creating those patches in the first place.
    :return (input, target): A tuple of tensors of input image and target image.
    """
    index = index // self.num_patches 
    img = Image.open(self.image_filenames[index]).convert('YCbCr') # Let's work with Grayscale images because they are simpler. To work with RGB images, simply use 'RGB'.  
    crop_size = self.crop_size
    img = patch_transform(crop_size)(img) # This function will be explained later.
    y, _,_ = img.split()
    input = y # We grab only the Y component of the image. This is the grayscale.
    target = input.copy() # We make a copy of the image.
    
    return transforms.ToTensor()(input), transforms.ToTensor()(target)

  
  blah blah ...
```

change 'YCbCr'
![image](https://user-images.githubusercontent.com/24697599/129065647-08100498-1f7f-4954-8cd9-9960e606a739.png)
to 'RGB'.

And also 
![image](https://user-images.githubusercontent.com/24697599/129065719-99c1bdcb-4848-4159-b94c-e4601bff3071.png)
make the input = img

Basically we pass in the colour image. You could probably refactor the dataset class such that it has a parameter called "colour" and if true do the above else if it is false extract out the grayscale stuff. But that's for refactoring. In essence make the change in the Dataset class and you should be good to go.

**Comment by dflateau** [comment]
Ah, right, I forgot about the data loader from the previous part of the other model. Thanks!  That seemed to work, and it's training!

**Comment by dflateau** [comment]
Do you remember how long the pre-trained RGB weights you posted were trained for? Were they trained in the same manner as the tutorial (just pure Meta-SRGAN?)

**Comment by pancakewaffles** [comment]
For the pretrained RGB weights, I am honestly not too sure. If their performance is not good, I think it’s better to just retrain on your own, especially now that you got the training loop working. Let me know if you want the DIV2K training data I used.

**Comment by dflateau** [comment]
Thanks for the quick response again. Yes, the results weren't great using those pretrained RGBweights (I hope I was doing inference correctly). The image I got very muich resembled the low resoltuion image I fed it, with more noise...

MIght try the div2k, which I have, and train that for a few days. When you used the div2k data, did you use the bicubic degradation option for the low res?

**Comment by pancakewaffles** [comment]
Yes I did. You may potentially get better scores than I did (which is good) as I did not do much hyperparameter tuning in the interest of time. So if you do get better scores that’s great, and not something to worry about.

**Comment by dflateau** [comment]
I've obtained really good results with single-scale models with EDSR + SRGAN fine-tuning. Would love to have the same thing but with arbitrary scales!

**Comment by dflateau** [comment]
Also, when you used the div2k data set - where's the best place to specify using the bicubic (or other drgraded) set of data? Or are you just degrading it yourself as you're doing in the tutorial?

**Comment by pancakewaffles** [comment]
Yes I decided to degrade on the spot because if I had to degrade for all possible scales it would take up a lot of space. Doing this dynamically and on the fly makes it a bit more efficient.

You can of course modify the TestDatasetFolder class to point your directory of low-res images. But then you would need a way to bring in the scale.

**Comment by dflateau** [comment]
Thanks - yeah, I forgot about the scale - my mind is still in single-scale mode. :) I'm in the middle training on the div2k set, so my notebook is occuipied - but a few questions, if you ahve the time:

1. am I going to save the weights/paramteres from the netG? Is that the correct model?
2. for inference, I'm guessing I'd do a netG(im_tensor,v), with the im_tensor the low-res image I'm processing, and v created from the size of that image?

**Comment by pancakewaffles** [comment]
Yep save it from the netG, and yes do a netG(im_tensor, v) for inference. But before that I think you need to set the scale so do a (for example) netG.set_scale(3.7) before passing in the im_tensor

**Comment by dflateau** [comment]
Right - I think it's netG.update_scale.  What would happen if I pass it something like netG.update_scale(3.762)? Would it interpolate the weights?

**Comment by pancakewaffles** [comment]
Yes it would update accordingly. What it really does is it scales the width to an int(original_width * 3.762) and height likewise. I don’t think there is an interpolation of weights.

**Comment by dflateau** [comment]
Ah, I see.  Thanks!

So, I've been watching this train for about 150 epochs, and... unfortunately, it's not converging on anything that looks like the high-res image. In fact, it almost seems like it's trying to optimize the model's output to look like the *low res* image, instead of the high res.  All the inferences  have it looking pretty close to that. Would that be a possiblity with the code that's there?

**Comment by pancakewaffles** [comment]
nah that’s how GANs train, at least I had similar experiences. What worked for me was a warm up period for the generator. So train the generator for like 10-15 epochs, then add in the discriminator for the remaining epochs.

Also try reducing the learning rate for the discriminator. It’s likely that the discriminator has “overpowered” the generator and so the generator isn’t learning much.

Maybe reduce the learning rate for discriminator by half.
